### DISCUSSION TITLE
What is provider.codeVerifier?

### DISCUSSION THREAD

**Discussion by Jerasaurus** [pocketbase/pocketbase#1775] [Q&A] [answered]
Hello! 

I'm trying to use SvelteKit to make an SSR app with google SSO as an option.

In the oauth example await every auth provider returned by `pb.collection('users').listAuthMethods()` has a codeVerifier parameter that is saved along with others to local storage.

Is saving auth_providers in localStorage the only way to accomplish this?

Furthermore, is it possible for the callback to be handled on the server as an API endpoint instead of by the client? 

Sorry for being new at this, and thanks in advance for any help!

**Comment by Jerasaurus** [comment]
Specifically I'm following along with this blog post https://haseebmajid.dev/posts/2022-12-20-how-to-setup-oauth2-with-sveltekit-and-pocketbase/

**Reply by edward3h** [reply]
I was following the same post today. I ran into an issue because `provider.codeVerifier` was not being saved between the login and redirect requests in the example code. I made it work by adding an extra cookie for codeVerifier.

**Reply by Jerasaurus** [reply]
I'm curious - in the post, however, the callback URL is a server endpoint that cant access any cookies. Did you change it so everything is handled by the client? (calling  `pb?.collection('users').authWithOAuth2(params)`)

**Reply by edward3h** [reply]
Yes, the callback code is '+server.ts', but it will receive cookies from the browser when it is called.
When you approve or whatever in the oauth provider page, it is telling the browser to go to the /callback URL, either by setting the window location, or with a redirect. Then the browser requests the Svelte server to handle that /callback URL, which will include the cookies in the request.

**Comment by ganigeorgiev** [comment, marked as answer]
> What is provider.codeVerifier?

`codeVerifier` is a random value that you send to the OAuth2 provider during the [PKCE](https://oauth.net/2/pkce/) flow.

Every parameter from the `listAuthMethods().authProviders` list can be generated client-side if you want. `state`, `codeVerifier`, `codeChallenge`, `authUrl`, etc. are pregenerated just for convenience.

> Is saving auth_providers in localStorage the only way to accomplish this?

You can use cookies or any other mechanism that can persist the `state` and `codeVerifier` parameter between the route with the "Login with ..." links and the redirect (aka. the actual OAuth2 login) handling.

> Furthermore, is it possible for the callback to be handled on the server as an API endpoint instead of by the client?

Yes, as long as you configure the OAuth2 provider callback url properly. Also keep in mind that browsers may have issues redirecting with cookies that have `SameSite=Strict` and you may have to change it to `SameSite=Lax` (see https://github.com/pocketbase/pocketbase/discussions/903#discussioncomment-4369424).

---

You can find a plain js OAuh2 integration example also in https://pocketbase.io/docs/authentication/#web-oauth2-integration.
_A simplified OAuth2 integration is already in the [roadmap](https://github.com/orgs/pocketbase/projects/2) and will be handled in the near future as part of #55._

**Reply by Jerasaurus** [reply]
Ah thank you! I'll give that a go later tonight
